README.md
Learning JavaScript Data Structures and Algorithms
====================================

[![Build Status](https://travis-ci.org/loiane/javascript-datastructures-algorithms.svg?branch=third-edition)](https://travis-ci.org/loiane/javascript-datastructures-algorithms)
[![codecov](https://codecov.io/gh/loiane/javascript-datastructures-algorithms/branch/third-edition/graph/badge.svg)](https://codecov.io/gh/loiane/javascript-datastructures-algorithms)

Source code of **Learning JavaScript Data Structures and Algorithms** book, third edition.

Work in Progress.

## List of available chapters:

* 01: [JavaScript, ECMAScript and TypeScript: a quick overview](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter01)
* 02: [Arrays](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter02)
* 03: [Stacks](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter03)
* 04: [Queues and Deques](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter04)
* 05: [LinkedLists](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter05)
* 06: [Sets](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter06)
* 07: [Dictionaries and Hashes](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter07)
* 08: [Recursion](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter08)
* 09: [Trees](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter09)
* 10: [Heap](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter10)
* 11: [Graphs](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter11)
* 12: [Sorting and Searching Algorithms](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter12)
* 13: [String and Math Algorithms](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter13)
* 14: [Algorithm Design and Techniques](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter14)
* 15: [Algorithm Complexity](https://github.com/loiane/javascript-datastructures-algorithms/tree/third-edition/examples/chapter15)

### Third Edition Updates

* Algorithms using ES2015+ (ES6+)
* New data structures and algorithms
* All chapters rewritten and reviewed 
* Three (3) new chapters
* Creation of a Data Structures and Algorithms library that can be used in the browser or with Node.js
* Algorithms tested with Mocha + Chai (test code available in `test` directory)
* **TypeScript** version of the source code included (library and tests)

## Project Structure

`src/js/index.js` file contains all the data structures and algorithms listed by chapter.

```
|_examples (how to use each data structure and algorithm, organized by chapter)
|_src 
|___js (source code: JavaScript version)
|_____data-structures
|_______models (classes used by DS: Node, ValuePair, ...)
|_____others (other algorithms such as palindome checker, hanoi tower)
|___ts (source code: TypeScript version)
|_____data-structures
|_______models
|_____others
|_test (unit tests with Mocha and Chai for src)
|___js (tests for JavaScript code)
|___ts (tests for TypeScript code)
```

## Installing and running the book examples With Node

* Install [Node](https://nodejs.org)
* Open terminal/cmd and change directoty to this project folder: `cd /Users/.../javascript-datastructures-algorithms` (Linux/Max) or `cd C:/.../javascript-datastructures-algorithms`
* run `npm install` to install all depencies
* To see the examples, run `http-server html` or `npm run serve`. Open your browser `http:\\localhost:8080` to see the book examples
* Or `cd html/chapter01` and run each javascript file with node: `node 02-Variables`

## Running the examples in the browser

* Right click on the html file you would like to see the examples, right click and 'Open with Chrome (or any other browser)'

* Or open the `examples/index.html` file to easily nagivate through all examples:

* Demo: [https://javascript-ds-algorithms-book.firebaseapp.com](https://javascript-ds-algorithms-book.firebaseapp.com)

<img src="examples/examples-screenshot.png">

Happy Coding!

## Other editions

| 1st edition   | 2nd edition   | 3rd edition   | 
| ------------- |:-------------:|:-------------:| 
| ![1st edition](https://d1ldz4te4covpm.cloudfront.net/sites/default/files/imagecache/ppv4_main_book_cover/4874OS_Learning%20JavaScript%20Data%20Structures%20and%20Algorithms.jpg)      | ![2nd edition](https://d255esdrn735hr.cloudfront.net/sites/default/files/imagecache/ppv4_main_book_cover/5493OS_5348_Learning%20JavaScript%20Data%20Structures%20and%20Algorithms,%20Second%20Edition.jpg) | ![2nd edition](https://d1ldz4te4covpm.cloudfront.net/sites/default/files/imagecache/ppv4_main_book_cover/B09377_MockupCover_0.png) |
| [Book link](http://amzn.to/1Y1OWPx)| [Book link](http://amzn.to/1TSkcA1)| [Book link](http://a.co/cbMlYmJ)|

Book link - first edition:
  - [Packt](https://www.packtpub.com/application-development/learning-javascript-data-structures-and-algorithms)
  - [Amazon](http://amzn.to/1Y1OWPx)
  - [Chinese version](http://www.ituring.com.cn/book/1613)
  - [Korean version](http://www.acornpub.co.kr/book/javascript-data-structure)

Book link - second edition:
 - [Packt](https://www.packtpub.com/web-development/learning-javascript-data-structures-and-algorithms-second-edition)
 - [Amazon](http://amzn.to/1TSkcA1)
 - [Chinese version](http://www.ituring.com.cn/book/2029)
 - [Brazilian Portuguese version](https://novatec.com.br/livros/estruturas-de-dados-algoritmos-em-javascript/)

 Book link - third edition:
 - [Packt](https://www.packtpub.com/web-development/learning-javascript-data-structures-and-algorithms-third-edition)
 - [Amazon](http://a.co/cbMlYmJ)

### Found an issue or have a question?

Please create an [Issue](https://github.com/loiane/javascript-datastructures-algorithms/issues) or [Pull Request](https://github.com/loiane/javascript-datastructures-algorithms/pulls) 


CHANGELOG.md
# Change Log
All notable changes to the "javascript-datastructures-algorithms" third edition source code bundle will be documented in this file.

Based on [Keep a Changelog](http://keepachangelog.com/).

## [Unreleased]
- Initial release


src/js/index.js
import * as _util from './util';

// chapters 05 and 07
export const util = _util;

// chapter 03
export { default as StackArray } from './data-structures/stack-array';
export { default as Stack } from './data-structures/stack';
export { default as hanoi } from './others/hanoi';
export { default as hanoiStack } from './others/hanoi';
export { default as baseConverter } from './others/base-converter';
export { default as decimalToBinary } from './others/base-converter';
export { default as parenthesesChecker } from './others/balanced-symbols';

// chapter 04
export { default as Queue } from './data-structures/queue';
export { default as Deque } from './data-structures/deque';
export { default as hotPotato } from './others/hot-potato';
export { default as palindromeChecker } from './others/palindrome-checker';

// chapter 05
export { default as LinkedList } from './data-structures/linked-list';
export { default as DoublyLinkedList } from './data-structures/doubly-linked-list';
export { default as CircularLinkedList } from './data-structures/circular-linked-list';
export { default as SortedLinkedList } from './data-structures/sorted-linked-list';
export { default as StackLinkedList } from './data-structures/stack-linked-list';

// chapter 06
export { default as Set } from './data-structures/set';

// chapter 07
export { default as Dictionary } from './data-structures/dictionary';
export { default as HashTable } from './data-structures/hash-table';
export { default as HashTableSeparateChaining } from './data-structures/hash-table-separate-chaining';
export { default as HashTableLinearProbing } from './data-structures/hash-table-linear-probing';
export { default as HashTableLinearProbingLazy } from './data-structures/hash-table-linear-probing-lazy';

// chapter 08
export { default as factorialIterative } from './others/factorial';
export { default as factorial } from './others/factorial';
export { default as fibonacci } from './others/fibonacci';
export { default as fibonacciIterative } from './others/fibonacci';
export { default as fibonacciMemoization } from './others/fibonacci';

// chapter 09
export { default as BinarySearchTree } from './data-structures/binary-search-tree';
export { default as AVLTree } from './data-structures/avl-tree';

// chapter 10
export { MinHeap } from './data-structures/heap';
export { MaxHeap } from './data-structures/heap';
export { default as heapSort } from './algorithms/sorting/heap-sort';

// chapter 11
export { default as Graph } from './data-structures/graph';
export { breadthFirstSearch } from './algorithms/graph/breadth-first-search';
export { BFS } from './algorithms/graph/breadth-first-search';
export { depthFirstSearch } from './algorithms/graph/depth-first-search';
export { DFS } from './algorithms/graph/depth-first-search';
export { dijkstra } from './algorithms/graph/dijkstra';
export { floydWarshall } from './algorithms/graph/floyd-warshall';
export { prim } from './algorithms/graph/prim';
export { kruskal } from './algorithms/graph/kruskal';

// chapter 12
export { shuffle } from './algorithms/shuffle/fisher–yates';

export { bubbleSort } from './algorithms/sorting/bubble-sort';
export { modifiedBubbleSort } from './algorithms/sorting/bubble-sort-improved';
export { bucketSort } from './algorithms/sorting/bucket-sort';
export { countingSort } from './algorithms/sorting/counting-sort';
export { insertionSort } from './algorithms/sorting/insertion-sort';
export { mergeSort } from './algorithms/sorting/merge-sort';
export { quickSort } from './algorithms/sorting/quicksort';
export { radixSort } from './algorithms/sorting/radix-sort';
export { selectionSort } from './algorithms/sorting/selection-sort';
export { shellSort } from './algorithms/sorting/shell-sort';

export { binarySearch } from './algorithms/search/binary-search';
export { interpolationSearch } from './algorithms/search/interpolation-search';
export { sequentialSearch } from './algorithms/search/sequential-search';
export { findMaxValue } from './algorithms/search/min-max-search';
export { findMinValue } from './algorithms/search/min-max-search';
